### ISSUE TITLE
Dev code for quality measure for generated data samples

### ISSUE THREAD

**Issue by mbaddar1** [mbaddar1/genmodel#18]
Objective to have some function q-measure(sample) that gives a measure on the generated sample, to use along with L2 loss while training the diffusion model. 
Dataset :
1. sklearn conc circles 
2.  "      half moon
3.  "      blob 

Focus mainly on the sinkhorn algorithm
https://github.com/fwilliams/scalable-pytorch-sinkhorn?tab=readme-ov-file 

misc links
https://gist.github.com/atabakd/ed0f7581f8510c8587bc2f41a094b518 
sinkhorn
https://amsword.medium.com/a-simple-introduction-on-sinkhorn-distances-d01a4ef4f085
https://papers.nips.cc/paper_files/paper/2013/hash/af21d0c97db2e27e13572cbf59eb343d-Abstract.html

**Comment by mbaddar1** [comment]
Started a script to experiment with sinkhorn 
https://github.com/mbaddar1/denoising-diffusion-pytorch/commit/ad5a5f4f7f4fd3547436f1c92eaf63f2fffb1a37
Data 
1. sklearn circles
2. MNIST zero and eight

The script with sinkhorn code is 
https://github.com/mbaddar1/denoising-diffusion-pytorch/blob/18-gen-data-quality-measure/stat_dist/sinkhorn.py
which is based on this repo
https://github.com/dfdazac/wassdistance
which I have forked here
https://github.com/dfdazac/wassdistance
and the illustration can be found here
https://github.com/dfdazac/wassdistance/blob/master/sinkhorn.ipynb


Results are
INFO:root:SH distances df = 
                       data_pair_name  sh_distance
0        sklearn_circles_same_params     0.075621
1  sklearn_circles_different_factors     0.149183
2   sklearn_circles_different_noises     0.648728
3                    mnist_two_zeros    21.578640
4                   mnist_two_eights    40.076790
5                   mnist_zero_eight   105.098251
which are quite intuitive : **To add illustration later**

**Comment by mbaddar1** [comment]
Understand more about the Wasserstein Distance thoery
https://alexhwilliams.info/itsneuronalblog/2020/10/09/optimal-transport/

**Comment by mbaddar1** [comment]
Back to the simple illustration in notebook
https://github.com/dfdazac/wassdistance/blob/master/sinkhorn.ipynb
Next step if to read 
Computational Optimal Transport
https://arxiv.org/pdf/1803.00567.pdf Ch. 4  Entropic Regularization of Optimal Transport

**Comment by mbaddar1** [comment]
Investigated the issue for normalizing Euclidean distance, this is what I found
https://www.cs.ucr.edu/~eamonn/Matrix_Profile_Tutorial_Part2.pdf 
https://www.youtube.com/watch?v=LnQneYvg84M&t=368s
also created another ticket to handle it #22

**Comment by mbaddar1** [comment]
Next steps to close this issue
at this line
https://github.com/mbaddar1/denoising-diffusion-pytorch/blob/18-gen-data-quality-measure/denoising_diffusion_pytorch/ddpm_sandbox/Trainer2.py#L189
insert a code to create a baseline for sinkhorn values for the dataset
and at this line
https://github.com/mbaddar1/denoising-diffusion-pytorch/blob/18-gen-data-quality-measure/denoising_diffusion_pytorch/ddpm_sandbox/Trainer2.py#L113
Add 
1. Sampling code
2. sinkhorn calculation
3. checkpoint saving 
4. meta data saving

**Comment by mbaddar1** [comment]
Added checkpoints to DDPM training loop
https://github.com/mbaddar1/denoising-diffusion-pytorch/pull/3/commits/2c71cdfca1a613065ea3a8385e6994b821740a81 
1. Apply sinkhorn eval each checkpoint
2. Save checkpoint model
3. Save checkpoint metadata

**Comment by mbaddar1** [comment]
next step , code snippet to load checkpoint models, generate samples from them and check visually. 
Also plot losses and sh scores
